### PR TITLE
Added common issue interface and implementation

### PR DESCRIPTION
**Pull request by duncanp-sonar**
* removed the daemon Issue type from all interfaces
* fixed up unit tests

Relates to #1470

### CODE REVIEW

File: src/Integration.Vsix/CFamily/CFamilyHelper.cs [hunk 1]
```diff
@@ -187,58 +187,58 @@ private static Request CreateRequest(FileConfig fileConfig, string absoluteFileP
...
-            return new Sonarlint.Issue()
-            {
-                FilePath = cfamilyIssue.Filename,
-                Message = cfamilyIssue.Text,
-                RuleKey = sqLanguage + ":" + cfamilyIssue.RuleKey,
-                Severity = Convert(defaultSeverity),
-                Type = Convert(defaultType),
-                StartLine = cfamilyIssue.Line,
-                EndLine = cfamilyIssue.EndLine,
+            return new AnalysisIssue
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/CFamily/CFamilyHelper.cs:190]: This is code that converted from a CFamily protobuf issue to a daemon protobuf issue. It now converts to an `IAnalysisIssue`.

File: src/Integration.Vsix/SonarLintDaemon/SonarLintDaemon.cs [hunk 2]
```diff
@@ -354,19 +356,75 @@ private bool IsVerbose()
...
+                startLine: daemonIssue.StartLine,
+                endLine: daemonIssue.EndLine,
+                startLineOffset: daemonIssue.StartLineOffset,
+                endLineOffset: daemonIssue.EndLineOffset
+                );
+
+        /// <summary>
+        /// Converts from the daemon issue severity enum to the standard AnalysisIssueSeverity
+        /// </summary>
+        internal /* for testing */ static AnalysisIssueSeverity Convert(DaemonIssueSeverity issueSeverity)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/SonarLintDaemon/SonarLintDaemon.cs:389]: `Convert` methods are in the same form as the corresponding CFamily ones.

File: src/Core/Analysis/IAnalysisIssue.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarLint.VisualStudio.Core
+{
+    public interface IAnalysisIssue
```
Comment (duncanp-sonar) [inline src/Core/Analysis/IAnalysisIssue.cs:23]: New analyzer-independent representation of issue, severity and issue type.

The property names are exactly the same as the daemon issue property names to minimise the changes to the existing code.

File: src/Integration.Vsix/SonarLintDaemon/DaemonIssueAdapter.cs [hunk 4]
```diff
@@ -28,15 +29,15 @@ internal class DaemonIssueAdapter : IFilterableIssue
         private readonly string wholeLineText;
         private readonly string lineHash;
 
-        public DaemonIssueAdapter(Sonarlint.Issue sonarLintIssue, string wholeLineText, string lineHash)
+        public DaemonIssueAdapter(IAnalysisIssue sonarLintIssue, string wholeLineText, string lineHash)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/SonarLintDaemon/DaemonIssueAdapter.cs:32]: This class is not longer specific to the daemon. It should be possible now to move it to `Core` and rename it.
There are few other interfaces/classes that can be moved. I'll do that in a separate PR.

File: src/Integration.Vsix.UnitTests/SonarLintDaemon/SonarLintDaemonTests.cs [hunk 5]
```diff
@@ -384,6 +388,40 @@ public void IsAnalyisSupported_AdditionalSupportIsActive()
             testableDaemon.IsAnalysisSupported(new[] { AnalysisLanguage.CFamily, AnalysisLanguage.Javascript }).Should().BeTrue();
         }
 
+        [TestMethod]
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/SonarLintDaemon/SonarLintDaemonTests.cs:391]: New tests for the enum conversions from daemon enum -> core enum.

File: src/Integration.Vsix.UnitTests/CFamily/CFamilyHelperTests.cs [hunk 6]
```diff
@@ -688,42 +687,55 @@ public void ToSonarLintIssue_SeverityAndTypeLookup()
...
             issue.RuleKey.Should().Be("lang1:rule3");
-            issue.Severity.Should().Be(Severity.Critical);
-            issue.Type.Should().Be(Sonarlint.Issue.Types.Type.Vulnerability);
+            issue.Severity.Should().Be(AnalysisIssueSeverity.Critical);
+            issue.Type.Should().Be(AnalysisIssueType.Vulnerability);
         }
 
         [TestMethod]
-        public void ConvertFromIssueSeverity()
+        [DataRow(IssueSeverity.Blocker, AnalysisIssueSeverity.Blocker)]
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/CFamily/CFamilyHelperTests.cs:703]: Miscellaneous: converted existing tests to be data-driven.

File: src/Integration.Vsix/SonarLintDaemon/ISonarLintDaemon.cs [hunk 7]
```diff
@@ -38,6 +38,6 @@ public interface ISonarLintDaemon : IAnalyzer, IDisposable
 
     public interface IIssueConsumer
     {
-        void Accept(string path, IEnumerable<Issue> issues);
+        void Accept(string path, IEnumerable<IAnalysisIssue> issues);
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/SonarLintDaemon/ISonarLintDaemon.cs:41]: It should now be possible to move this interface to `Core`.

**Comment by duncanp-sonar** [comment]
This change affects quite a lot of files, but most of the changes are just changing variable types.
I've commented the substantive changes.

File: src/TestInfrastructure/DummyAnalysisIssue.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.Core;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
+{
+    // Properties are settable to simplify creating test instances
+    public class DummyAnalysisIssue : IAnalysisIssue
```
Comment (rita-g-sonarsource) [inline src/TestInfrastructure/DummyAnalysisIssue.cs:26]: Do we really need this? We can use Mock<IAnalysisIssue>

**Pull request merged**